include/onnxruntime/core/graph/graph.h

```diff
@@ -780,6 +783,12 @@ class Graph {
   */
   void SetInputs(const std::vector<const NodeArg*>& inputs);
 
+  /** Explicitly update specified graph initializers.
+  @param updated_input_map An map contains the name of the initializer that needs to update and the
+  corresponding new value. This function will overwrite the old NodeArg with the new value.
+  */
+  void SyncWithFP16Initializer(const std::unordered_map<std::string, NodeArg*>& updated_input_map);
+
```

Comment:
Could this be done by a helper using RemoveInitializedTensor and AddInitializedTensor instead of adding a very specific method to Graph API?

Also note that those methods do a number of things to keep name_to_initial_tensor_ in sync, which I would have expected to still be needed.

Reply:
hmm, the problem here is that, graph object has two fields to track initializers and they are not in sync. `name_to_initial_tensor_`  tracks tensorProto and `graph_inputs_including_initializers_` tracks NodeArg. `RemoveInitializedTensor` and `AddInitializedTensor` only update `name_to_initial_tensor_`, but not `graph_inputs_including_initializers_`. `TransformGraphForMixedPrecision` currently only call `AddInitializedTensor` to just update `name_to_initial_tensor_`, this causes a bug later in a different function which assumes `graph_inputs_including_initializers_` is up-to-date. 

To fix that, this method is added to keep both in-sync. Agree this is not ideal. `RemoveInitializedTensor` and `AddInitializedTensor` should check and ensure the two fields are in sync. But it will require a signature change to make that happen, which is beyond the scope of this PR. If this approach seems reasonable to you, I can create a task to re-visit it later. Thoughts?






---
In reply to: [434262216](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434262216) [](ancestors = 434262216)

Reply:
I think fixing Remove/AddInitializedTensor to correctly update graph_inputs_including_initializers would be the right thing to do. The graph data structure should be self-consistent.

---
In reply to: [434792198](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434792198) [](ancestors = 434792198,434262216)

Reply:
I think there's a misunderstanding. graph_inputs_including_initializers_ pairs with graph_inputs_excluding_initializers_. These are lists of required and optional graph inputs, where an 'optional' graph input is backed by an initializer with a default value.

name_to_initial_tensor_ is the complete list of all initializers, including those that do not have a matching graph input. 

Unless you're adding a new graph input so the initializer can be overridden, you should not be updating graph_inputs_including_initializers_.

---
In reply to: [434893890](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434893890) [](ancestors = 434893890,434792198,434262216)

Reply:
I agree with you that this should be the ideal way to fix this bug. The two fields should be in sync and updated in one place. Given that these functions are so widely used, and the change requires a signature change, in order to pass in both tensorProto and NodeArg, it would touch a lot of files. There are also common cases when `AddInitializedTensor` is called, but the corresponding NodeArg is not yet created. So I think it makes sense to create a task to track it and seperate this to a different PR. 

---
In reply to: [434893890](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434893890) [](ancestors = 434893890,434792198,434262216)

Reply:
>Unless you're adding a new graph input so the initializer can be overridden, you should not be updating graph_inputs_including_initializers_.
`TransformGraphForMixedPrecision()` creates new fp16 initializers input, and replace the existing fp32 ones by changing the corresponding consumer nodes' NodeArgs. I guess this qualifies as a new default initializer and in this case, `graph_inputs_including_initializers_` should be updated?

---
In reply to: [434905541](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434905541) [](ancestors = 434905541,434893890,434792198,434262216)

Reply:
Synced with Scott offline to get more clarification of the concept. 

FYI here if any one is interested: `name_to_initial_tensor_` is designated for graph embeded initializers that exclude any newly created internal ones. `graph_inputs_including_initializers_` is equal to `graph_inputs` which is paired with `graph_inputs_excluding_initializers_`. The former specifies graph inputs including initializers that user can optionally override with; while the latter includes inputs without any initializer, which means the inputs user has to provide data for. For this case, as the newly created tensor is solely internal, `name_to_initial_tensor_` should keep intact. 

---
In reply to: [434905748](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434905748) [](ancestors = 434905748,434893890,434792198,434262216)